net/multi_listen.go

```diff
@@ -0,0 +1,151 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+)
+
+// connAndErrPair packs results of accept call on actual listeners together along with the
+// index of the listener. It is used for communication between main and listener goroutines.
+type connAndErrPair struct {
+	index int
+	conn  net.Conn
+	err   error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	latestAcceptedIndex atomic.Int32
+	addrs               []net.Addr
+	listeners           []net.Listener
+
+	wg           sync.WaitGroup
+	connAndErrCh chan connAndErrPair
+	stopCh       chan struct{}
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(addresses []string) (net.Listener, error) {
+	return multiListen(addresses, net.ResolveTCPAddr, net.Listen)
+}
+
+func multiListen(
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connAndErrCh: make(chan connAndErrPair),
+		stopCh:       make(chan struct{}),
+	}
+
+	for _, address := range addresses {
+		addr, err := resolveTCPAddrFunc("tcp", address)
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		var network string
+		host, _, err := net.SplitHostPort(addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+		switch IPFamilyOf(ParseIPSloppy(host)) {
+		case IPv4:
+			network = "tcp4"
+		case IPv6:
+			network = "tcp6"
+		default:
+			// close all listeners
+			_ = ml.Close()
+			return nil, fmt.Errorf("failed to identify ip family of address '%s", addr.String())
+		}
+
+		l, err := listenFunc(network, addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		ml.addrs = append(ml.addrs, addr)
+		ml.listeners = append(ml.listeners, l)
+	}
+
+	for i := range ml.listeners {
+		index := i
+		ml.wg.Add(1)
+		// spawn a go routine for every listener to wait for incoming connection requests
+		go func() {
+			defer ml.wg.Done()
+			for {
+				conn, err := ml.listeners[index].Accept()
+				if err != nil {
+					select {
+					// Accept() will throw "use of closed network connection" when listener is closed,
+					// we can ignore that error and break out of this goroutine.
+					case <-ml.stopCh:
+						return
+					default:
+						ml.connAndErrCh <- connAndErrPair{conn: conn, err: err, index: index}
+					}
+				}
+				ml.connAndErrCh <- connAndErrPair{conn: conn, err: err, index: index}
+			}
```

Comment:
I think you can remove the stopCh just using `connAndErrCh` 
```suggestion
				select {
				case ml.connAndErrCh <- connAndErrPair{conn: conn, err: err, index: index}
				default:
				         return
			}
```

Reply:
There can be some corner cases in that approach, if there is a genuine error in Accept() call (when listeners are not closed) then the worker go-routine will try consuming from `connAndErrCh` and if the channel already has something to read from then that connection will be dropped.

Reply:
don't completely follow, is an unbuffered channel and now you always send anything through the channel expect when the stopCh is closed, what is the different on using the stopCh to sync the goroutines vs using the connAndErrCh?

Reply:
accept() immediately fails with "use of closed network connection" when the listener is closed and I used stopCh to determine that and exit the goroutine. I can use atomic.Bool instead.

This error is not exported by stdlib, so I need some way to determine the exit condition for the goroutines.

Reply:
The atomic bool doesn't prevent the race at all. The accept go-routine may be scheduled out _right before_:
```go
				ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
```
_then_ Close() is called, and the go-routine is re-scheduled, and the send will be on a closed channel -> panic().

IMHO you should keep it simple. Close the channel _after_ `ml.wg.Wait()`, and just send all errors from Accept() on the channel. Let the user sort it out. They should be prepared that Accept() returns an error if the listener is closed (and will likely ignore them)

Reply:
> The atomic bool doesn't prevent the race at all. The accept go-routine may be scheduled out right before:
> then Close() is called, and the go-routine is re-scheduled, and the send will be on a closed channel -> panic().

When **Close()** is called it sets the atomic bool to true, and in the accept go-routine I simply exit/return when atomic bool is true instead of writing on the channel and no panic would occur.
```go
			for {
				conn, err := ml.listeners[index].Accept()
				if err != nil && ml.closed.Load() {
					return
				}
				ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
			}
```



> IMHO you should keep it simple. Close the channel after ml.wg.Wait(), and just send all errors from Accept() on the channel. Let the user sort it out. They should be prepared that Accept() returns an error if the listener is closed (and will likely ignore them)

In that case I would need to exit the accept for loop whenever any error occurs, right? Other wise the go-routine would never end and **ml.Accept()** would be blocked on **ml.wg.Wait()** forever. 
But we should only exit on "use of closed network connection" which occurs when listener is closed, right? Other errors are genuine and should be passed to user.

Right now this atomic bool is being used to determine the exit condition.

Reply:
```go
			for {
				conn, err := ml.listeners[index].Accept()
				if err != nil && ml.closed.Load() {
					return
				}
// if the go-routine is suspended *right here*, after reading the "closed" boolean, but before sending on the
// channel, and *then* another go-routine calling close and thus set the "closed" boolean (which is already read
// as "true" by this go-routine) and closing the channel, then you will get a panic() when this go-routine is
// re-scheduled and sends on a closed channel! This is true regardless if the boolean is atomic, or not.
				ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
			}
```

Reply:
> In that case I would need to exit the accept for loop whenever any error occurs, right?

Um, yes you are right. If you use a "closed" boolean (which should not be atomic), you must use a Mutex, e.g.
```go
			for {
				conn, err := ml.listeners[index].Accept()
				ml.mutex.Lock()
				if ml.closed {
					ml.mutex.Unlock()
					return
				} else {
					ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
				}
				ml.mutex.Unlock()
			}
// And
func (ml *multiListener) Close() error {
	ml.mutex.Lock()
	ml.closed = true
	close(ml.connErrIndexCh)
	ml.mutex.Unlock()
	for i := range ml.listeners {
		_ = ml.listeners[i].Close()
	}
	ml.wg.Wait()
	return nil
}
```

Reply:
how a mutex instead of atmoic.Bool ?
```go
			for {
			       conn, err := ml.listeners[index].Accept()
			       ml.mux.Lock()
			       if err != nil && ml.closed {
					return
			       }
			       ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
			       ml.mux.Unlock()
			}
```

Reply:
The Accept() is "semantically weird". Since the multi-listener _always_ have Accept's running on all addresses, stale entries will linger on the channel until Accept() is called the next time. For instance if the multi-listener is created with one IPv4 and one IPv6 (the use-case this implementation is mainly addressing), and the client is using [happy eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs) and send a connect on both families in parallel, then the first will succeed and the runner-up will linger in the channel until Accept() is called again (which may be much later)

Reply:
In https://github.com/kubernetes/utils/pull/311#discussion_r1641776790: do not check err != nil, if ml.closed==true you must never write to the channel (it's closed), and unlock before return.

Reply:
> The Accept() is "semantically weird".

Excellent callout, Lars.  Is it better to be semantically close but not identical, or to emphasize the semantic difference?  It would require more app changes, but we could do something like:

```
// I am not sure we need the Listener here
acceptHandler := func(listener net.Listener, conn net.Conn, err error) {
    if err != nil {
        // log it
    }
    // handle an accepted socket however I need to, maybe a channel
    // or some other queue or handler-pool mechanism
}
ml := knet.MultiListen("tcp", acceptHandler, "10.9.8.7:80", "10.1.2.3:8000")
```